src/censored.jl

```diff
@@ -0,0 +1,409 @@
+    if !iszero(prob_lower)
+        v += prob_lower * abs2(lower - μ)
+    end
+    if !iszero(prob_upper)
+        v += prob_upper * abs2(upper - μ)
+    end
+    return v
+end
+
+# this expectation also uses the following relation:
+# 𝔼_{x ~ τ}[-log d(x)] = S[τ] - log P_{x ~ d₀}(l ≤ x ≤ u)
+#   + (P_{x ~ d₀}(x = l) (log P_{x ~ d₀}(x = l) - log P_{x ~ d₀}(x ≤ l)) + 
+#      P_{x ~ d₀}(x = u) (log P_{x ~ d₀}(x = u) - log P_{x ~ d₀}(x ≥ u))
+#   ) / P_{x ~ d₀}(l ≤ x ≤ u),
+# where S[τ] is the entropy of τ.
+
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    if lower !== missing
+        prob_lower_inc = cdf(d0, lower)
+        pl = value_support(typeof(d0)) === Discrete ? pdf(d0, lower) : 0
+        prob_lower = prob_lower_inc - pl
+        entropy_lower = -xlogx(prob_lower_inc)
+    else
+        pl = prob_lower = entropy_lower = 0
+    end
+    if upper !== missing
+        prob_upper = ccdf(d0, upper)
+        pu = value_support(typeof(d0)) === Discrete ? pdf(d0, upper) : 0
+        entropy_upper = -xlogx(prob_upper + pu)
+    else
+        pu = prob_upper = entropy_upper = 0
+    end
+    result = entropy_lower + entropy_upper
+    prob_interval = 1 - (prob_lower + prob_upper)
+    # truncation contains no probability
+    iszero(prob_interval) && return result
+
+    dtrunc = _to_truncated(d)
+    result += prob_interval * entropy(dtrunc) - xlogx(prob_interval) + xlogx(pl) + xlogx(pu)
+    return result
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = cdf(d0, x)
+        _eqnotmissing(x, upper) ? one(result) : result
+    elseif _eqnotmissing(x, upper)
+        _ccdf_inc(d0, x)
+    else
+        result = pdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : zero(result)
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = logcdf(d0, x)
+        _eqnotmissing(x, upper) ? zero(result) : result
+    elseif _eqnotmissing(x, upper)
+        _logccdf_inc(d0, x)
+    else
+        result = logpdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : oftype(result, -Inf)
+    end
+
+end
+
+function loglikelihood(d::Censored, x::AbstractArray{<:Real, M}) where {M}
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    log_prob_lower = lower === missing ? 0 : logcdf(d0, lower)
+    log_prob_upper = upper === missing ? 0 : _logccdf_inc(d0, upper)
+
+    return sum(x) do xi
+        _in_open_interval(xi, lower, upper) && return logpdf(d0, xi)
+        _eqnotmissing(xi, lower) && return log_prob_lower
+        _eqnotmissing(xi, upper) && return log_prob_upper
+        T = float(Base.promote_eltype(log_prob_lower, log_prob_upper))
+        return T(-Inf)
```

Comment:
This seems to be type unstable. E.g., if `lower = 0.0f0`, `upper = 10.0f0`, `x = [-2.0, 1.0]`, `d0 = Normal{Int}(0, 1)`. Then `log_prob_lower isa Float32`, `log_prob_upper isa Float32` and hence for the first observation `x[1]` the function evaluates to `-Inf32` whereas for the second observation `x[2]` it evaluates to `logpdf(d0, 1.0) isa Float64`.